src/events/attributes.rs

```diff
@@ -437,12 +438,139 @@ impl<'a> Iterator for Attributes<'a> {
                 self.position = j;
                 return attr!(start_key..end_key, i..j);
             }
-            Some((i, _)) => err!(Error::UnquotedValue(i)),
+            Some((i, _)) => err!(AttrError::UnquotedValue(i)),
             None => return attr!(start_key..end_key),
         }
     }
 }
 
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+/// Errors that can be raised during parsing attributes.
+///
+/// Recovery position in examples shows the position from which parsing of the
+/// next attribute will be attempted.
+#[derive(Debug, PartialEq)]
+pub enum AttrError {
+    /// Attribute key not followed by with `=`, position relative to start of
+    /// owning tag is provided.
+    ///
+    /// Example of input that raises this error:
+    ///
+    /// ```xml
+    /// <tag key another="attribute"/>
+    /// <!--     ^~~ error position, recovery position (8) -->
+    /// ```
+    ///
+    /// That error can be raised only if iterator in XML mode.
+    ExpectedEq(usize),
+    /// Attribute value not found after `=`, position relative to start of owning
+    /// tag is provided.
+    ///
+    /// Example of input that raises this error:
+    ///
+    /// ```xml
+    /// <tag key = />
+    /// <!--       ^~~ error position, recovery position (10) -->
+    /// ```
+    ///
+    /// Notice, that this error can be returned only for the last attribute in
+    /// the list, because otherwise any content after `=` will be threated as
+    /// a value. The XML
+    ///
+    /// ```xml
+    /// <tag key = another-key = "value"/>
+    /// <!--                   ^ ^- recovery position (24) -->
+    /// <!--                   '~~ error position (22) -->
+    /// ```
+    ///
+    /// will be threated as `Attribute { key = b"key", value = b"another-key" }`
+    /// and or [`Attribute`] is returned, or [`AttrError::UnquotedValue`] is raised,
+    /// depending on the parsing mode.
+    ExpectedValue(usize),
+    /// Attribute value not quoted, position relative to start of owning tag
+    /// is provided.
+    ///
+    /// Example of input that raises this error:
+    ///
+    /// ```xml
+    /// <tag key = value />
+    /// <!--       ^    ^~~ recovery position (15) -->
+    /// <!--       '~~ error position (10) -->
+    /// ```
+    ///
+    /// That error can be raised only if iterator in XML mode.
+    UnquotedValue(usize),
+    /// Attribute value not finished with a matching quote, position relative to
+    /// start of owning tag and a quote is provided. That position is always a last
```

Comment:
I understand this to mean that "the position will be after a quote character" but to say that the position is relative to the "start of the owning tag and a quote" is a little bit confusing.  It sounds like it could mean either of

A) the position will be on a quote character, which isn't true
B) the position is relative to both, which isn't possible

I think "and a quote" could be removed and it would still make sense (given the examples).

Reply:
Feel free to rephrase it and I'll add it to the other your suggestions.